alvr/sockets/src/stream_socket/mod.rs

```diff
@@ -0,0 +1,397 @@
+        port: u16,
+        max_packet_size: usize,
+        timeout: Duration,
+    ) -> ConResult<StreamSocket> {
+        let (send_socket, receive_socket) = match self {
+            StreamSocketBuilder::Udp(socket) => {
+                udp::connect(&socket, server_ip, port, timeout).to_con()?;
+                udp::split_multiplexed(socket, max_packet_size).to_con()?
+            }
+            StreamSocketBuilder::Tcp(listener) => {
+                let socket = tcp::accept_from_server(&listener, Some(server_ip), timeout)?;
+                tcp::split_multiplexed(socket, timeout).to_con()?
+            }
+        };
+
+        Ok(StreamSocket {
+            send_socket: Arc::new(Mutex::new(send_socket)),
+            receive_socket,
+            queues: HashMap::new(),
+        })
+    }
+
+    #[allow(clippy::too_many_arguments)]
+    pub fn connect_to_client(
+        timeout: Duration,
+        client_ip: IpAddr,
+        port: u16,
+        protocol: SocketProtocol,
+        dscp: Option<DscpTos>,
+        send_buffer_bytes: SocketBufferSize,
+        recv_buffer_bytes: SocketBufferSize,
+        max_packet_size: usize,
+    ) -> ConResult<StreamSocket> {
+        let (send_socket, receive_socket) = match protocol {
+            SocketProtocol::Udp => {
+                let socket =
+                    udp::bind(port, dscp, send_buffer_bytes, recv_buffer_bytes).to_con()?;
+                udp::connect(&socket, client_ip, port, timeout).to_con()?;
+                udp::split_multiplexed(socket, max_packet_size).to_con()?
+            }
+            SocketProtocol::Tcp => {
+                let socket = tcp::connect_to_client(
+                    timeout,
+                    &[client_ip],
+                    port,
+                    send_buffer_bytes,
+                    recv_buffer_bytes,
+                )?;
+                tcp::split_multiplexed(socket, timeout).to_con()?
+            }
+        };
+
+        Ok(StreamSocket {
+            send_socket: Arc::new(Mutex::new(send_socket)),
+            receive_socket,
+            queues: HashMap::new(),
+        })
+    }
+}
+
+pub struct StreamSocket {
+    send_socket: Arc<Mutex<Box<dyn MultiplexedSocketWriter + Send>>>,
+    receive_socket: Box<dyn MultiplexedSocketReader + Send>,
+    queues: HashMap<u16, StreamRecvQueues>,
+}
+
+impl StreamSocket {
+    pub fn request_stream<T>(&self, stream_id: u16) -> StreamSender<T> {
+        StreamSender {
+            inner: Arc::clone(&self.send_socket),
+            stream_id,
+            payload_offset: self.send_socket.lock().payload_offset(),
+            next_packet_index: 0,
+            used_buffers: vec![],
+            _phantom: PhantomData,
+        }
+    }
+
+    // max_concurrent_buffers: number of buffers allocated by this call which will be reused to
+    // receive packets for this stream ID. If packets are not read fast enough, the shards received
+    // for this particular stream will be discarded
```

Comment:
Honestly don't understand why that's needed, especially the way it seems to be implemented with the queue. We don't really care about the number of buffers currently in use by the code using the received data, missing deallocs there are just a regular memory leak. Wouldn't the packet receiver itself just end up discarding stuff because no one's calling receive to actually take the data?

Reply:
I think now we have more safety nets (dropping the connection if not producing frames for 1 second) but for correctness, each component (sockets in this case) should be well behaved in all circumstances. Unit tests would probably be beneficial